KEYCLOAK-11019 Initial support for lazy offline user-session loading

thomasdarimont opened the pull request:
This PR introduces initial support for lazy-loading of offline user-sessions.
See the discussion on the [keycloak-dev group](https://groups.google.com/g/keycloak-dev/c/ef0P_pj8ndI/m/ObBrtCjCDAAJ)

As a start, we allow to disable the pre-loading of offline sessions from the database. At a later stage we can also disable the pre-loading from a remote-cache storage.

If an offline user-session is missing from the infinispan sessions or clientSession cache, we try to find the offline user-session in the database. If the session is found, we dynamically import it into the appropriate infinispan cache.

Additionally we now allow to disable loading of offline user-sessions during startup via the new `preloadOfflineSessionsFromDatabase` property of the `infinispan userSessions` provider.

patch-standalone-ha.cli:
```bash
embed-server --server-config=standalone-ha.xml --std-out=echo

/subsystem=logging/console-handler=CONSOLE:write-attribute(name=level,value=DEBUG)

if (outcome == failed) of /subsystem=logging/logger=org.keycloak.models.sessions.infinispan/:read-resource
  /subsystem=logging/logger=org.keycloak.models.sessions.infinispan:add(level=DEBUG)
end-if

echo SETUP: Customize Keycloak UserSessions SPI configuration
# Add dedicated userSession config element to allow configuring elements.
if (outcome == failed) of /subsystem=keycloak-server/spi=userSessions/:read-resource
  echo SETUP: Add missing userSessions SPI
  /subsystem=keycloak-server/spi=userSessions:add()
  echo
end-if

echo SETUP: Configure built-in "infinispan"  UserSessions loader
if (outcome == failed) of /subsystem=keycloak-server/spi=userSessions/provider=infinispan/:read-resource
  echo SETUP: Add missing "infinispan" provider
  /subsystem=keycloak-server/spi=userSessions/provider=infinispan:add(enabled=true)
  /subsystem=keycloak-server/spi=userSessions/provider=infinispan:write-attribute(name=properties.preloadOfflineSessionsFromDatabase,value=${env.KEYCLOAK_INFINISPAN_SESSIONS_PRELOAD_DATABASE:false})
  /subsystem=keycloak-server/spi=userSessions:write-attribute(name=default-provider,value=infinispan)
  echo
end-if
```

build the keycloak distribution from source:
```
mvn -Pdistribution -pl distribution/server-dist -am -Dmaven.test.skip clean install
```
once this is finished, you can find your Keycloak distribution in `distribution/server-dist/target` as .zip or .tar.gz.

Apply `patch-standalone-ha.cli` to `standalone-ha.xml`
```
echo "yes" | bin/jboss-cli.sh --file patch-standalone-ha.cli
```

Start Keycloak:
```
bin/standalone.sh -c standalone-ha.xml --debug -Dwildfly.statistics-enabled=true
```

You should now see a DEBUG log message like:
```
...
13:09:30,850 DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory] (ServerService Thread Pool -- 65) Skipping pre-loading of userSessions from persistent storage
...
```

If you try to obtain a new access token for a given (offline) refresh-token you should see the following in the log:
```
...
13:09:37,538 DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProvider] (default task-1) Offline user-session not found in infinispan, attempting UserSessionPersisterProvider lookup for sessionId=9c5aa5fa-1611-4c70-a4ec-10bb0edf5690
13:09:37,555 DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProvider] (default task-1) Offline user-session found in UserSessionPersisterProvider, attempting to reimport user-session for sessionId=9c5aa5fa-1611-4c70-a4ec-10bb0edf5690
13:09:37,573 DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProvider] (default task-1) Offline user-session imported, trying another lookup for sessionId=9c5aa5fa-1611-4c70-a4ec-10bb0edf5690
13:09:37,574 DEBUG [org.keycloak.models.sessions.infinispan.InfinispanUserSessionProvider] (default task-1) Offline user-session found after import for sessionId=9c5aa5fa-1611-4c70-a4ec-10bb0edf5690
...
```

thomasdarimont committed:
KEYCLOAK-11019 InfinispanUserSessionProvider should fallback to database query if in-memory lookup fails for offline sessions

Allow to configure pre-loading of offline user-sessions

We now allow to configure whether offline user-sessions should be
preloaded on startup from the database.

We now trigger offline user-sessions loading when
UserSessionProvider#getOfflineUserSessionsStream(realm,user) is called.

This ensures that the user can revoke the offline_access consent
via the account console. Additionally the (offline) user consents
are now displayed correctly for a user in the admin-console.

Offline user-sessions are now correctly display if offline user-sessions
are loaded lazily.

Add note about unsupported lazy offline session loading by broker info:
Currently we don't support lazy offline user-session lookup by
brokerSessionId and brokerUserId.

See: https://github.com/keycloak/keycloak/pull/7722#pullrequestreview-578669985

Keep behavior of 'logout all sessions' and don't delete dangling offline sessions.
As discussed in https://github.com/keycloak/keycloak/pull/7722#discussion_r566989120

Throw on unsupported dynamic offline user-session lookups from database

pedroigor commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java:
```diff
@@ -239,6 +241,52 @@ public void removeExpired(RealmModel realm) {
+
+        PersistentUserSessionAdapter userSession = result.get(0);
+
+        TypedQuery<PersistentClientSessionEntity> clientSessionQuery = em.createNamedQuery("findClientSessionsByUserSessions", PersistentClientSessionEntity.class);
```
Isn't possible to reuse the same code from `loadUserSessions` method?

thomasdarimont replied:
Thanks for your feedback @pedroigor 

> org.keycloak.models.jpa.session.JpaUserSessionPersisterProvider#loadUserSessions
That's what I tried at first, but it turned out that, the session referenced by the `lastUserSessionId` parmeter is not included in the result... that's why I decided to implement the new method.

thomasdarimont replied:
... of course we could change the `sess.userSessionId > :lastSessionId` condition in the findUserSessions JPA query to `sess.userSessionId >= :lastSessionId`, but I'd also need to adjust the other conditions or pass dummy values (0?) for `lastCreatedOn`.

hmmm, I still think that the two methods are fine for now.

pedroigor reviewed (changes requested):
Overall, looks good to me.

It should not break existing behavior while making possible to lazy load offline sessions. The impact on sessions list should be expected as we discussed in the mailing list.

Let's see what @mposolda thinks about it.

thomasdarimont commented:
Rebased on current master

martin-kanis reviewed (changes requested):
@thomasdarimont Thank you for the PR. 
Is this feature somehow tested? If not, maybe it's worth to add some tests.

Next the `JpaUserSessionPersisterProvider.java` should use Java Streams as we recently changed collections to streams where ever that make sense. See the suggestions inline. 

Finally I would add javadocs to newly introduced methods in `UserSessionPersisterProvider.java` interface.

martin-kanis commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java:
```diff
@@ -239,6 +241,52 @@ public void removeExpired(RealmModel realm) {
+        TypedQuery<PersistentClientSessionEntity> clientSessionQuery = em.createNamedQuery("findClientSessionsByUserSessions", PersistentClientSessionEntity.class);
+        clientSessionQuery.setParameter("userSessionIds", Collections.singleton(userSessionId));
+        clientSessionQuery.setParameter("offline", offlineStr);
+        List<PersistentClientSessionEntity> clientSessions = clientSessionQuery.getResultList();
```
Also consider using streams here:
```suggestion
        closing(clientSessionQuery.getResultStream()).forEach(clientSession -> {
            PersistentAuthenticatedClientSessionAdapter clientSessAdapter = toAdapter(userSession.getRealm(), userSession, clientSession);
            Map<String, AuthenticatedClientSessionModel> currentClientSessions = userSession.getAuthenticatedClientSessions();

            // Case when client was removed in the meantime
            if (clientSessAdapter.getClient() == null) {
                removedClientUUIDs.add(clientSession.getClientId());
            } else {
                currentClientSessions.put(clientSession.getClientId(), clientSessAdapter);
            }
        });
```

thomasdarimont replied:
I adapted the closing(..) part, but kept the rest as is, see [my comment](https://github.com/keycloak/keycloak/pull/7722#issuecomment-763155458) below

thomasdarimont commented:
@martin-kanis Thanks or your suggestions. I'd prefer to leave the code as is for now, to keep the style in this class consistent. 
A "streamification" should IMHO be applied afterwards in a holistic refactoring.

Btw. I think the JPA query result stream in  `org.keycloak.models.jpa.session.JpaUserSessionPersisterProvider#loadUserSessions` should also be wrapped in `org.keycloak.utils.StreamsUtil#closing`.

I'm fine with adding the javadoc comments to the new API.

Regarding tests, how can I test offline-sessions with a custom InfinispanUserSessionProviderFactory configuration?

thomasdarimont commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java:
```diff
@@ -239,6 +241,52 @@ public void removeExpired(RealmModel realm) {
+        userSessionQuery.setParameter("userSessionId", userSessionId);
+        userSessionQuery.setMaxResults(1);
+
+        List<PersistentUserSessionAdapter> result = userSessionQuery.getResultStream()
```
I adapted the closing(..) part, but kept the rest as is, see [my comment](https://github.com/keycloak/keycloak/pull/7722#issuecomment-763155458) below

martin-kanis commented:
@thomasdarimont Actually the [streamification epic](https://issues.redhat.com/browse/KEYCLOAK-14011) was finished by [this commit](https://github.com/keycloak/keycloak/commit/9f580e3ed868223a7d5c3d59481e5f9535a9e163) which also covers the `JpaUserSessionPersisterProvider.java` class. For this reason we need all new code where it makes sense to use Java streams.

thomasdarimont commented:
@martin-kanis Thanks for the hint. I just rebased the PR to the current master and revised the Streams API usage as you suggested.

Do you have an idea how I could address the testing issues I mentioned above?

pedroigor commented:
> @thomasdarimont Actually the [streamification epic](https://issues.redhat.com/browse/KEYCLOAK-14011) was finished by [this commit](https://github.com/keycloak/keycloak/commit/9f580e3ed868223a7d5c3d59481e5f9535a9e163) which also covers the `JpaUserSessionPersisterProvider.java` class. For this reason, we need all new code where it makes sense, to be using Java streams.

I also think "streamification" should not be a design principle, but used when (after testing and profiling) and where it makes sense.

stianst commented:
I don't really know what was agreed on in the thread as it was a lot of back and forth, mixing short-term and long-term. This is kinda why we have the design proposals in the first place so we can have a written down agreed upon plan, as mailing lists are okay for discussions, but terrible way of capturing the output of a discussion.

So as lazy loading from the DB is disabled by default, I assume that's because it's breaking something, as otherwise it would just make no sense to have this as an option and just always lazy load offline sessions from the DB. What specifically is being broken?

stianst reviewed (commented):
I don't really know what was agreed on in the thread as it was a lot of back and forth, mixing short-term and long-term. This is kinda why we have the design proposals in the first place so we can have a written down agreed upon plan, as mailing lists are okay for discussions, but terrible way of capturing the output of a discussion.

So as lazy loading from the DB is disabled by default, I assume that's because it's breaking something, as otherwise it would just make no sense to have this as an option and just always lazy load offline sessions from the DB. What specifically is being broken?

stianst reviewed (changes requested):
I'm going to request at least some changes here:

* Describe what the side-effects are. It seems like to me that this may break anything that views/lists offline sessions in cases where the sessions hasn't been lazy loaded yet, including the ability for users and admins to view/manage offline sessions. If that's the case I don't really like the idea of merging this PR
* If it's an option it needs documenting. If there are going to be difference in behaviour it has to be documented. As it stands this is a "hidden" config option that only a select few will know about, and anyone that stumbles across it won't know what benefits/drawbacks the option provides

pedroigor commented:
I would say that loading data and blocking at boot time is by itself a broken design. And for this particular case, it has a huge impact on use cases using offline sessions.

We have quite a few workarounds already to overcome this problem (e.g.: sessionsPerSegment, not sure if documented) that works until a certain point.

IMO, the proposed changes provide a good balance for those facing the problem we are fixing while sacrificing some functionality that is not used or important. It should also help to pave the road for further improvements and in a proper fix that not necessarily impacts existing functionality.

thomasdarimont commented:
Hello @stianst, @pedroigor, @martin-kanis,

thanks a lot for the great discussion and useful feedback!

# Regarding the question "Describe what the side-effects are.":

1) With the changes from this PR and `preloadOfflineSessionsFromDatabase=true` (current default)  
a) Keycloak will pre-load offline sessions at startup as usual - which might block the server start for large number of offline-sessions for multiple minutes  
b) The number of offline-sessions per client in the admin-console is based on the offline-user session count in memory  
c) (*new*) If an offline user session is not found in memory, a database lookup is performed - if successful the offline-session is imported into infinispan.  
This useful in Kubernetes rolling upgrade, that's prone to "loose" some offline-sessions due to incomplete infinispan cache synchronizations.  
d) Everything else (session listing in admin-console etc.) works as before  
e) Admin-Console / Manage / Session -> will show the number of Active Sessions from memory and number of Offline-sessions from memory  

2) With the changes from this PR and `preloadOfflineSessionsFromDatabase=false`  
a) Keycloak will NOT pre-load offline sessions at startup -> user-session / client-session caches will be empty at (server/cluster) start  
b) The number of offline-sessions per client in the admin-console is based on the offline-user session count in the database  
c) If an offline user session is not found in memory, a database lookup is performed - if successful the offline-session is imported into infinispan.  
This useful in Kubernetes rolling upgrade, that's prone to "loose" some offline-sessions due to incomplete infinispan cache synchronizations.  
d) The client session listing in the admin-console might not show all offline-sessions or even be empty if no offline session were loaded yet.  
As there might be houndreds or millions of offline sessions it is IMHO not useful at all to be able to list all via the admin-console.
The warning "Warning, this is a potentially expensive operation depending on the number of offline tokens." gives a hint at this as well.  
e) Admin-Console / Manage / Session -> will show the number of Active Sessions from memory and number of Offline-sessions from the database  

# Regarding "If it's an option it needs documenting"
I'd document this alongside the mentioned `sessionsPerSegment`. As I also couldn't find a section about this in the docs,
I'd recommend to add a section "Performance Tuning" to [Server Installation and Configuration Guide](https://www.keycloak.org/docs/latest/server_installation/index.html).

@stianst This PR is a IMHO workaround that solves a current flaw in Keycloak's offline-session handling that affects a large number of Keycloak deployments. This helps to solve scalability problems for IMHO a large number of users today.

# Some useful enhancements
- Offline user-sessions caches could be configured with expiration to remove unused cache items after some time to save memory  
- Listing of offline-sessions for a client in the admin-console could be improved with a user search  

This is of course not a replacement for the mid-long term storage architecture redesign :)

stianst commented:
@thomasdarimont Thanks. Main thing I'd want is that there is no limitations that prevents people from being able to use it. For the account console will it load offline sessions from the DB for the given user so the user can see offline sessions even though they haven't been pre-loaded yet? For the admin console if I look at sessions for a given user does it show them then?

Reminding myself how the sessions management looks like in the admin console there's options to logout sessions for a given user, or all sessions, not for an individual application. So as long as we can logout for a given user I think we're covered here as a "working workaround" ;)

thomasdarimont commented:
@stianst thanks for the quick reply.
> For the account console will it load offline sessions from the DB for the given user so the user can see offline sessions even though they haven't been pre-loaded yet?
Good catch. If the offline user-session pre-loading is disabled and the offline user-session is not loaded yet, then the "revoke grant" button is missing in the applications page of the account-console. The [current implementation of the ApplicationsBean](https://github.com/keycloak/keycloak/blob/99a70267d9b68da9462eaee82ba2d6f2ae4fae54/services/src/main/java/org/keycloak/forms/account/freemarker/model/ApplicationsBean.java#L53) seems to load the offline-sessions only from the cache. 
I think this should be changed to try a database lookup as well, in case the offline session was not found in the cache.
We could either import the offline session or just extract the metadata from the database, which might be the proper way if we say that only a token refresh request should trigger offline user-session loading.

> For the admin console if I look at sessions for a given user does it show them then?
If the offline user-session is not loaded yet, then user/consents will not list the offline_access grant. 
Solution could be the same as above, we could load only the from the database without importing the offline-session.

I give this a quick try now.

thomasdarimont commented:
@stianst @pedroigor  I just updated the PR with the changes to fix the issues mentioned by stian
- The application page in the (old) account-console now triggers offline user-session loading and shows the "revoke access" button.
- The consents listing of the user page in the admin-console now triggers offline user-session loading and shows the offline-access consent

I think the only caveat rest is that admin-console -> clients -> client -> Offline Access will only show offline-sessions that are loaded in memory. The (total) number of offline sessions will be the correct count of offline-sessions in the database.

thomasdarimont commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -270,6 +336,14 @@ private UserSessionEntity getUserSessionEntity(RealmModel realm, String id, bool
 
+    private Stream<UserSessionModel> getUserSessionsFromPersistenceProviderStream(RealmModel realm, UserModel user, boolean offline) {
+        UserSessionPersisterProvider persister = session.getProvider(UserSessionPersisterProvider.class);
+        return persister.loadUserSessionsStream(0, Integer.MAX_VALUE, realm, user, offline)
```
I think it is good to have pagination support here, but I wasn't sure about default limits.

mposolda replied:
IMO pagination support may not be so urgent for the users? I suppose that it will be very rare that single user has for example 1000  offline sessions? Pagination makes sense for the clients IMO and it is already available in the admin console (however not yet supported in the PR as mentioned in the comment).

thomasdarimont commented:
rebased and updated to latest master

mposolda commented:
@thomasdarimont Thanks for the PR and sorry for the late review.

This is good work, but my vote is to make sure that Keycloak functionality works correctly and there are no side-effects on various places for the case that sessions are not yet loaded from DB to infinispan. Like for example missing logout or incorrect counts in admin console etc. Otherwise, we can count with some more bug reports in the future (including possible security issues). And from the review, it looks that there are few places missing.

It seems that those calls to UserSessionProvider missing the calls to DB and hence they can return incorrect results and miss the sessions:

1.
```
Map<String, Long> getActiveClientSessionStats(RealmModel realm, boolean offline);
```
This one is used for example by admin console when you go to "Sessions" tab on the left. Incorrect results here can have security side-effects (EG. In case of leaked realm, admin has an impression that there are no offline sessions, even if they are and hence he may miss invalidating them etc)

2.
```
Stream<UserSessionModel> getOfflineUserSessionsStream(RealmModel realm, ClientModel client, Integer firstResult, Integer maxResults);
```
This one is used by the admin console when you go to Clients -> someClient -> OFfline Access -> Click "Show offline access" . The admin console already supports pagination in the UI and API also supports pagination. The tooltip with the "Warning", which you mentioned in the previous comment, is incorrect as the pagination is there

3.
```
Stream<UserSessionModel> getOfflineUserSessionByBrokerUserIdStream(RealmModel realm, String brokerUserId);
```
It seems this one can have side-effects for security and missing logouts.

The first 2 issues can be addressed by using DB as the "authoritative storage" instead of infinispan cache. Similarly like you are doing for the users.

The issue 3 (searching by brokerUserId), I am not sure but it may not be easily possible in the persister though. In this case we can probably search the user by brokerAlias+brokerId with UserProvider, which is possible (See JpaUserProvider.searchForUserStream and searching by UserModel.IDP_ALIAS and UserModel.IDP_USER_ID). And then search persister by returned UserModel, which you already added in this PR.


Another thing are missing tests. Since the parameter "preloadOfflineSessionsFromDatabase" is true by default, there is no automated test of the new functionality at all. It would be ideal if there is automated test for all the use-cases and REST endpoints above to make sure that correct results are returned.

thomasdarimont commented:
Thanks for the great feedback @mposolda !

I'll try to address 1) and 2) today. For 3) and the tests I'd need some guidance.

Would you be available for a quick google meet session in the afternoon?

thomasdarimont commented:
@mposolda 
Regarding 1) Admin-Console -> Sessions
This uses `InfinispanUserSessionProvider#getActiveClientSessionStats` which currently fetches the information from the infinispan cache. If we say that `getActiveClientSessionStats` only returns the current (offline) sessions that are actually in-use, then we could even say that the current behaviour is correct and does not need to be changed. 
Btw. the offline session count shown in Admin-Console -> Clients -> Client -> Offline Access would show the total number of offline sessions. 

I can change this to also load the offline user-session counts from the database, if you prefer that.

Regarding 2)  Clients -> someClient -> Offline Access -> Click "Show offline access" 
Working on it...

thomasdarimont commented:
@mposolda I still need some input about 1) Admin-Console -> Sessions stats and testing

Regarding 2) Clients -> someClient -> Offline Access -> Click "Show offline access"
"Show offline user-sessions" for a client in account-console works now as expected with the latest commit.

hmlnarik reviewed (changes requested):
Thanks @thomasdarimont for the PR! The idea is overall good, though there might be some conflict resolution needed since there are also #7738 and #7736, both touching the same files and cleaning the code a bit.

I share the view of @mposolda's and @stianst, the offline token as when loaded from persistence store needs to behave transparently just the same as if it was original infinispan implementation. Especially revoking the offline tokens seems a tough issue which relates to your question:

> Is it save to use the given brokerUserId as a userId for an offline_user_session table lookup?

Sadly, no, `brokerUserId` is stored as JSON encoded in the `OFFLINE_USER_SESSION.DATA` column. The same applies to broker session id.

I believe this would only be possible to overcome by altering the database structure and migration at this moment. Since we 
aim to be no-downtime-friendly, that would mean that there needs to be some redundancy at least for some time, and the database columns would only be added, and the `data` column would not be altered. The broker session and user would be then stored both in the `data` as it is now and in the new two columns. Then it would be possible to follow your suggestion in the code.

Apart from that I have few minor comments to the code, please find those inline.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/models/session/UserSessionPersisterProvider.java:
```diff
@@ -56,6 +56,37 @@
+     * @param offline
+     * @return
+     */
+    Stream<UserSessionModel> loadUserSessionsStream(Integer firstResult, Integer maxResults, RealmModel realm, UserModel user, boolean offline);
```
I'll adapt.

thomasdarimont replied:
Should be fixed now.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/models/session/UserSessionPersisterProvider.java:
```diff
@@ -56,6 +56,37 @@
+     * @param offline
+     * @return
+     */
+    Stream<UserSessionModel> loadUserSessionsStream(Integer firstResult, Integer maxResults, RealmModel realm, ClientModel client, boolean offline);
```
I'll adapt.

thomasdarimont replied:
Fixed.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/models/session/UserSessionPersisterProvider.java:
```diff
@@ -78,6 +109,19 @@
+     * @param clientModel
+     * @return
+     */
+    int getUserSessionsCount(boolean offline, ClientModel clientModel);
```
I'll adapt and generally agree, but I think the realm makes no sense here since, the sole real implementation queries the `PersistentClientSessionEntity` via underlying query `PersistentClientSessionEntity.findClientSessionsCountByClient` filtered by the given client id. Also the entity `PersistentClientSessionEntity` does not have a `realmId` attribute.

I can add the parameter but I'd need to ignore it in the JPA implementation.

thomasdarimont replied:
Fixed, by adding the `RealmModel` parameter which is ignored by the JPA implementation.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/models/session/UserSessionPersisterProvider.java:
```diff
@@ -56,6 +56,37 @@
+     * @param offline
+     * @return
+     */
+    UserSessionModel loadUserSession(String userSessionId, boolean offline);
```
I'll adapt. Although the userSessionId is pretty much unique already.

thomasdarimont replied:
Fixed.

thomasdarimont commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java:
```diff
@@ -40,8 +40,14 @@
-
+                " order by sess.createdOn,sess.userSessionId"),
+        @NamedQuery(name="findUserSession", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
+                " AND sess.userSessionId = :userSessionId"),
```
I'll adapt.

thomasdarimont replied:
This should be fixed.

thomasdarimont commented:
@mposolda @hmlnarik I adapted the PR according to your suggestions.

Apart from the IdP brokering stuff, the rest should be working fine.
Would it be possible to run the full test-suite with `preloadOfflineSessionsFromDatabase=false` ?

mposolda commented:
@thomasdarimont @hmlnarik Few points to the issue 3.

3.a) Regarding DB structure, I see that DATA column currently has un-encoded JSON. Something like this: `{"ipAddress":"127.0.0.1","authMethod":"openid-connect","rememberMe":false,...`
Maybe there is a way to just use SQL, which will have some condition similar to `DATA like '%"brokerUserId":":brokerUserId"%` or something like this? This is not so great solution and not 100% sure it will work, but maybe yes...


3.b) If the previous is not a good option, you can search UserProvider by "brokerUserId" . The snippet you pointed won't work as "brokerUserId" is something different than userId. Our JpaUserProvider has some support for searching by Broker alias and broker user Id. I have this pseudo-snippet just to illustrate the idea :
```
        if (offline && loadOfflineSessionsStatsFromDatabase) {

	    // Extract the brokerAlias and idpUserId from the argument passed to the UserSessionProvider.getOfflineUserSessionByBrokerUserIdStream .
            // I assume the format based on looking at the LogoutEndpoint.backchannelLogoutFederatedUserId() method
            String[] parts = brokerUserId.split("\.");
            String idpAlias = parts[0];
            String idpUserId = parts[1];

	    // Search user provider now
            Map userSearch = new HashMap();
            userSearch.put(UserModel.IDP_ALIAS, idpAlias);
            userSearch.put(UserModel.IDP_USER_ID, idpUserId);
            Stream<UserModel> users = session.users().searchForUsers(realm, userSearch);

// I think there is single user matching (not 100% sure, will be good to doublecheck)
            Optional<UserModel> firstUser = users.findFirst();

// Check if user exists
            if (firstUser.exists()) {
               // Now just search the sessions based on the existing user
               return getOfflineUserSessions(realm, firstUser.get()); 
            } else {
               return nullOrEmptyStream();
            }                        
        }
```
This approach has the issue that it possibly won't work for the custom userStorage providers. But it should work for JPA based on looking at JpaUserProvider.searchForUsersStream.

mposolda commented:
@thomasdarimont Regarding test, it is somehow possible to run whole testsuite with "preloadOfflineSessionsFromDatabase=false", but not so easily. And from the long-term perspective, we will need to:
- test this "regularly" as part of normal testsuite runs
- At the same time, we probably can't introduce another "matrix" to our pipeline (or github actions) for every configuration option like that

So my vote is to just create another separate test class and add some testing scenarios to that class. This class will just use "preloadOfflineSessionsFromDatabase=false" . As an example you can take a look at FixedHostnameTest and it's superclass AbstractHostnameTest . AbstractHostnameTest currently contains some logic to switch hostname providers on the fly.

I think that for long-term, we may need to have some better solution then "if" with few conditions and rather have some abstraction in the testsuite, which will allow switch providers and eventually their configuration for each runtime server (including quarkus). But that is likely out of scope of this PR. So hopefully if we can have some hack similar to AbstractHostnameTest.reset(), it may be ok for this PR in my opinion :-)

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java:
```diff
@@ -37,11 +37,27 @@
         @NamedQuery(name="updateUserSessionLastSessionRefresh", query="update PersistentUserSessionEntity sess set lastSessionRefresh = :lastSessionRefresh where sess.realmId = :realmId" +
                 " AND sess.offline = :offline AND sess.userSessionId IN (:userSessionIds)"),
-        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline"),
+        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline and (sess.realmId = :realmId or :realmId is null)"),
```
Is there any valid way to have `realmId == NULL`? Any session belongs to a realm, so if `realmId` is `null`, this would indicate a bug.
```suggestion
        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline and sess.realmId = :realmId"),
```

This seems to stem from the change in `getUserSessionsCount(boolean offline)` that now delegates to `getUserSessionsCount(realm, offline)` and passes in `null` in `realm`. This change should be reverted, as the condition `realmID IS null` should lead to empty result.

The change is only needed in the case of the newly introduced `int getUserSessionsCount(boolean offline, ClientModel clientModel);` variant that needs to have `RealmModel` as well.

thomasdarimont replied:
realm==null means that the count of all user sessions is returned. This is currently used by [OfflinePersistentUserSessionLoader](https://github.com/keycloak/keycloak/blob/8432513daa1c19437b48b2a87dcc97e093212008/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/OfflinePersistentUserSessionLoader.java#L67).

hmlnarik replied:
That's correct, but the query this way can effectively cause a database to perform evaluation of the `OR` condition instead of using index lookup. For that reason, two separate queries for this two cases would be preferrable.

thomasdarimont replied:
Okay, I'll check that and add a dedicated query if necessary.

thomasdarimont replied:
I did some tests with `and sess.realmId = coalesce(:realmId, sess.realmId)` it seems to always use the proper index.

hmlnarik reviewed (changes requested):
I am not convinced that the logout without properly logging out brokered sessions is sufficient. Effectively this means that:
- From the perspective of the user, user is apparently logged out on keycloak. Yet the session still remains active.
- Hence the respective offline session, when queried, will reappear, since it remains in the database.

> So my vote is to just create another separate test class and add some testing scenarios to that class. This class will just use "preloadOfflineSessionsFromDatabase=false" .

Agree.

> 3.a. [...] Maybe there is a way to just use SQL, which will have some condition similar to DATA like '%"brokerUserId":":brokerUserId"% or something like this? This is not so great solution and not 100% sure it will work, but maybe yes...

This option leads to performance bottleneck, given that this would be a substring search on an unindexed column, so I don't think this path should be followed.

> 3.b) If the previous is not a good option, you can search UserProvider by "brokerUserId" .

That would be a viable option.

Is logout by broker session ID covered?

mposolda commented:
@hmlnarik Good catch. It looks that logout by broker session ID is also not covered and it's something, which I missed in the initial review. Unfortunately for this one, I don't see any good option as (3.b) probably won't help here. List of some not good options:
1) Use (3.a) approach and live with the fact that performance of this won't be good in case of preloading of offline sessions disabled. We will need to document it.
2) Do something to make sure that DB searching is improved. TBH this seems to me like potential big cost and risk as I guess this may require to do some re-structure of DB columns etc.
3) Just document that backchannelLogout with offline tokens requires preloading to be enabled for now - and counting that we will solve in the new store. To be defensive, we can throw the exception from InfinispanUserSessionProvider in case that attempt to do search by brokerSessionId is done and preloading is dsiabled

One thing to consider - it looks that backchannel logout of broker offline sessions requires to use the parameter "revoke_offline_access" to be added to the "events" claim into the OIDC logout token. However I don't see this parameter in any OIDC/OAuth2 specification and even in our documentation.
@pedroigor Do you know if "revoke_offline_access" comes from some specification or if it is "special" and undocumented feature specific to Keycloak OIDC only?

If it is custom undocumented thing, we can most likely go with (3) as there is probably single user using this feature (Guy who implemented it) :-)

hmlnarik commented:
> @hmlnarik Good catch. It looks that logout by broker session ID is also not covered and it's something, which I missed in the initial review. Unfortunately for this one, I don't see any good option as (3.b) probably won't help here. List of some not good options:
> 
> 1. Use (3.a) approach and live with the fact that performance of this won't be good in case of preloading of offline sessions disabled. We will need to document it.

I believe this effectively means sequential scanning of the offline sessions table upon each such logout, and significant part of this PR is trying to avoid exactly this sequential scan that is performed now only once upon start.

> 2. Do something to make sure that DB searching is improved. TBH this seems to me like potential big cost and risk as I guess this may require to do some re-structure of DB columns etc.

I see this as a viable option, some details are sketched in [the last paragraph above](https://github.com/keycloak/keycloak/pull/7722#pullrequestreview-578669985).

> 3. Just document that backchannelLogout with offline tokens requires preloading to be enabled for now - and counting that we will solve in the new store. To be defensive, we can throw the exception from InfinispanUserSessionProvider in case that attempt to do search by brokerSessionId is done and preloading is disabled

This is yet another option worth exploring. If there would be a big fat warning in the documentation with the explanation of that option and an error logged every time unsupported logout occurs, I see it as acceptable.

mposolda commented:
@hmlnarik I see. My vote is to go with the option (3) assuming that the feature of "revoke_offline_access" is really only Keycloak specific thing, which is not even documented right now. Otherwise go with (2) and add the column including migration support. Will wait for @pedroigor input as well regarding this...

thomasdarimont commented:
Should offline user-sessions be removed on admin-console->sessions -> logout all?
The infinispan code path seems to do this in [InfinispanUserSessionProvider#removeUserSessions(RealmModel)](https://github.com/keycloak/keycloak/blob/c0beca7744a5eea3858ba9881a222fcfd8a64802/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java#L609)

Here I'd need to remove all offline sessions from the database to, right? Because after the sessions have been logged out, the offline sessions doesn't work anymore, even if `preloadOfflineSessionsFromDatabase=true`. I think this is a bug in the logic. Either logout all sessions should spare offline tokens OR we need to delete the offline sessions in the database too, otherwise they will end up accumulating loads of zombi offline sessions in the database.

pedroigor commented:
@mposolda In regards to the `revoke_offline_access` parameter. The spec does not define now offline_access should be handled but as possible improvements/investigations on how to do it properly. I think that is due to the fact that uses cases for logout are slightly different than those related to offline sessions.

So, that parameter is ours and should indicate that offline sessions are revoked on a per-id or all basis. Per-id should be fine if you are using offline as an "online" session. So I think it should be fine to go for #3.

And that is one of my points about this PR, any collateral effect should be expected as a trade-off for loading sessions on a per ID basis. As well as the implications of doing it in terms of user revocation. For some use cases, users won't need to deal directly with offline sessions but online sessions.

thomasdarimont commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -607,6 +722,11 @@ public void removeUserSessions(RealmModel realm) {
                 RemoveUserSessionsEvent.createEvent(RemoveUserSessionsEvent.class, InfinispanUserSessionProviderFactory.REMOVE_USER_SESSIONS_EVENT, session, realm.getId(), true),
                 ClusterProvider.DCNotify.LOCAL_DC_ONLY);
+
+        log.debugf("Removing user sessions for realm via persistence provider. realm=%s", realm.getId());
```
`RealmAdminResource#logoutAll` called when admin-console -> "logout all sessions" is invoked.

The infinispan code path invalidates and clears all sessions from memory, hence offline-tokens can no longer be used.
However, the offline user sessions remain listed in the database, but they cannot be reclaimed / reused anymore. Even a sever restart doesn't bring them life.

I think this dangling (offline) user-sessions needs to be cleared from the database, if the sessions are removed.

mposolda replied:
It seems to me that `RealmAdminResource#logoutAll` doesn't logout offline sessions in current master? I see this based on some testing and also when looking at the code. The `REMOVE_USER_SESSIONS_EVENT` registered in the InfinispanUserSessionsProviderFactory calls the `provider.onRemoveUserSessionsEvent(sessionEvent.getRealmId())` and that one seems to remove just "online" sessions?

The behaviour of "Logout all" in the admin console is not very clear. It is not possible to differentiate between "Logout all online and offline" sessions or "Logout just all online sessions" etc. I think we have some JIRAs to improve that. But my vote is, that this PR focus just on the offline sessions preloading and doesn't change any other existing behaviour and hence keep "Logout all" to still focus just on online sessions.

thomasdarimont replied:
Okay, thanks for clarification. I'll remove the offline-session removal from the "logout all" call.

thomasdarimont replied:
May I leave the code `UserSessionPersisterProvider#removeUserSessions(realm,offline)` in? 
This could then at least be used by users in custom extensions who need this functionality.
Btw. I know of several customers who need a true logout-all feature including offline user-sessions.

mposolda commented:
@pedroigor Thanks! So I take I see that "revoke_offline_sessions" is our parameter, which is not documented and hence probably not used by anyone despite the author of backchannel PR.

@hmlnarik Based on the info from above, my vote is to just document this and don't change the DB as backchnannel logout of offline broker sessions is corner case use-case, not documented, not yet added in RHSSO 7.4 (Was added in August 2020), and probably not used by anyone despite the contributor of the feature. So my vote is to go with (3) rather than risk of DB related issues. But you know better how much is the overhead of adding another DB column (or columns if we want to add both brokerSessionId and brokerUserId, which would be likely better than rely on searching userProvider as that is another overhead), so I am rather leaving the decision to you.

hmlnarik commented:
> @hmlnarik Based on the info from above, my vote is to just document this and don't change the DB as backchnannel logout of offline broker sessions is corner case use-case, not documented, not yet added in RHSSO 7.4 (Was added in August 2020), and probably not used by anyone despite the contributor of the feature. So my vote is to go with (3) rather than risk of DB related issues. But you know better how much is the overhead of adding another DB column (or columns if we want to add both brokerSessionId and brokerUserId, which would be likely better than rely on searching userProvider as that is another overhead), so I am rather leaving the decision to you.

I agree, option (3) looks like a safest option we have on the table, i.e. _Just document that backchannelLogout with offline tokens requires preloading to be enabled for now_ and _[...] throw the exception from InfinispanUserSessionProvider in case that attempt to do search by brokerSessionId is done and preloading is disabled_.

thomasdarimont commented:
@hmlnarik 
```
I agree, option (3) looks like a safest option we have on the table, i.e. Just document that backchannelLogout with offline tokens requires preloading to be enabled for now and [...] throw the exception from InfinispanUserSessionProvider in case that attempt to do search by brokerSessionId is done and preloading is disabled.
```
If I understood you right, then I'd need to throw an exception here: 
https://github.com/keycloak/keycloak/pull/7722/files#diff-1445188baf4e2f8d90de96fc76d0423dd86425d2a0a9ede5c3e20d88c3f85077R360
Which exception should I throw?

thomasdarimont commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -270,8 +336,37 @@ private UserSessionEntity getUserSessionEntity(RealmModel realm, String id, bool
+                return persister.loadUserSessionsStream(realm, user, offline, 0, null);
+            }
+
+            // TODO add support for offline user-session lookup by brokerSessionId
```
@hmlnarik Based on the [discussion here](https://github.com/keycloak/keycloak/pull/7722#issuecomment-770723908): 
which exception should I throw here?

mposolda replied:
My vote is to throw ModelException as InfinispanUserSessionProvider is model implementation.

I am not 100% sure if we want to catch that exception in LogoutEndpoint and transform into some proper ErrorResponseException, or if we just don't handle exception? But IMO this is not user-related error (like for example updating password which doesn't match the password policy), but it is rather system error. Administrator should be aware that the use-case is not supported, so it will be ideal to make sure that it is shown in the server log and probably should be passed to the Event API as well.

Will be great if you have a chance to add automated test for this error scenario or at least check it manually. WDYT?

And I would throw the exception for both getOfflineUserSessionByBrokerSessionId and getOfflineUserSessionByBrokerUserId.

thomasdarimont replied:
Thanks for your input @mposolda !
I just adapted the appropriate method to throw the ModelException.

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentClientSessionEntity.java:
```diff
@@ -30,15 +30,16 @@
         @NamedQuery(name="findClientSessionsByUserSession", query="select sess from PersistentClientSessionEntity sess where sess.userSessionId=:userSessionId and sess.offline = :offline"),
-        @NamedQuery(name="findClientSessionsByUserSessions", query="select sess from PersistentClientSessionEntity sess where sess.offline = :offline and sess.userSessionId IN (:userSessionIds) order by sess.userSessionId")
+        @NamedQuery(name="findClientSessionsByUserSessions", query="select sess from PersistentClientSessionEntity sess where sess.offline = :offline and sess.userSessionId IN (:userSessionIds) order by sess.userSessionId"),
+        @NamedQuery(name="findClientSessionsCountByClient", query="select count(sess) from PersistentClientSessionEntity sess where sess.offline = :offline and sess.clientId = :clientId")
```
This query should be backed by index.

thomasdarimont replied:
The query `findClientSessionsCountByClient` requires a full table scan for the generated query:
```sql
-- PersistentClientSessionEntity.findClientSessionsCountByClient
select count(persistent0_.CLIENT_ID) as col_0_0_
from OFFLINE_CLIENT_SESSION persistent0_
where persistent0_.OFFLINE_FLAG='1'
  and persistent0_.CLIENT_ID='dce1b319-cac8-41af-bf11-123b768060c1'
 ```
![image](https://user-images.githubusercontent.com/314690/107160038-13036d00-6994-11eb-88e1-b9684349ccf4.png)

With the index:
```sql
create index IDX_OFFLINE_CSS_PRELOAD on offline_client_session (offline_flag, client_id);
```
Created via 
```xml
    <changeSet author="keycloak" id="13.0.0-KEYCLOAK-11019">
        <createIndex tableName="OFFLINE_CLIENT_SESSION" indexName="IDX_OFFLINE_CSS_PRELOAD">
            <column name="OFFLINE_FLAG" type="VARCHAR(4)"/>
            <column name="CLIENT_ID" type="VARCHAR(36)"/>
        </createIndex>
    </changeSet>
```
The index is used and the query is much faster.
![image](https://user-images.githubusercontent.com/314690/107160040-1a2a7b00-6994-11eb-8b3f-58bfbe28f318.png)

mposolda replied:
@thomasdarimont Thanks for the changes and adding the tests .

@hmlnarik Does the index introduced in the commit https://github.com/keycloak/keycloak/pull/7722/commits/02b6e62e1614118957f7a7b5ccea1a5f06c2ee55 work for you? Could it have side-effect on the performance when introducing index?

hmlnarik replied:
The indices could be improved, see below. The analysis of how to improve the indices is [here](https://docs.google.com/spreadsheets/d/14nk2i7bG0dCeI0MLqANoxEXX1VIehyfOZ5qFQIXoCrs/edit?usp=sharing)

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java:
```diff
@@ -32,16 +32,32 @@
  */
 @NamedQueries({
-        @NamedQuery(name="deleteUserSessionsByRealm", query="delete from PersistentUserSessionEntity sess where sess.realmId = :realmId"),
+        @NamedQuery(name="deleteUserSessionsByRealm", query="delete from PersistentUserSessionEntity sess where sess.realmId = :realmId and sess.offline = coalesce(:offline, sess.offline) "),
```
This leads currently to seq scan of the table.

thomasdarimont replied:
The query `deleteUserSessionsByRealm` results in the following SQL:
```sql
delete from OFFLINE_USER_SESSION sess
where sess.realm_id = 'acme'
  and sess.offline_flag = coalesce(cast(null as text), sess.offline_flag)
```
Without modifications this results in a full table scan as you mentioned.
![image](https://user-images.githubusercontent.com/314690/107160595-835fbd80-6997-11eb-871a-d865b56134c3.png)

With the following index:
```
<createIndex tableName="OFFLINE_USER_SESSION" indexName="IDX_OFFLINE_USS_CLEANUP">
            <column name="OFFLINE_FLAG" type="VARCHAR(4)"/>
            <column name="realm_id" type="VARCHAR(36)"/>
        </createIndex>
```
the SQL statement uses the index and is much faster:
![image](https://user-images.githubusercontent.com/314690/107160624-a7230380-6997-11eb-9774-25b7be54cede.png)

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java:
```diff
@@ -32,16 +32,32 @@
                 " AND (sess.createdOn > :lastCreatedOn OR (sess.createdOn = :lastCreatedOn AND sess.userSessionId > :lastSessionId))" +
-                " order by sess.createdOn,sess.userSessionId")
-
+                " order by sess.createdOn,sess.userSessionId"),
+        @NamedQuery(name="findUserSession", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
+                " AND sess.userSessionId = :userSessionId AND sess.realmId = :realmId"),
+        @NamedQuery(name="findUserSessionsByUserId", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
+                " AND sess.realmId = :realmId AND sess.userId = :userId"),
+        @NamedQuery(name="findUserSessionsByClientId", query="SELECT sess FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
+                " ON sess.userSessionId = clientSess.userSessionId AND clientSess.clientId = :clientId WHERE sess.offline = :offline " +
+                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId AND clientSess.clientId = :clientId"),
+        @NamedQuery(name="findUserSessionsCountsByClientId", query="SELECT clientSess.clientId, count(clientSess) " +
+                " FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
+                " ON sess.userSessionId = clientSess.userSessionId " +
+                // find all available offline user-session for all or specific clients in a realm
+                " AND clientSess.clientId = coalesce(:clientId, clientSess.clientId) " +
+                " WHERE sess.offline = :offline " +
+                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId " +
+                // find all available offline user-session for all or specific clients in a realm
+                " AND clientSess.clientId = coalesce(:clientId, clientSess.clientId) " +
+                " GROUP BY clientSess.clientId")
```
Most of these queries resort to plain seq scan because of missing / inappropriate indices.

thomasdarimont replied:
The query `findUserSession` results in the following SQL:
```sql
--- PersistentUserSessionEntity.findUserSession

select persistent0_.OFFLINE_FLAG         as OFFLINE_1_41_,
       persistent0_.USER_SESSION_ID      as USER_SES2_41_,
       persistent0_.CREATED_ON           as CREATED_3_41_,
       persistent0_.DATA                 as DATA4_41_,
       persistent0_.LAST_SESSION_REFRESH as LAST_SES5_41_,
       persistent0_.REALM_ID             as REALM_ID6_41_,
       persistent0_.USER_ID              as USER_ID7_41_
from OFFLINE_USER_SESSION persistent0_
where persistent0_.OFFLINE_FLAG='1'
  and persistent0_.USER_SESSION_ID='400c53b3-1c21-40a5-8a40-727f1f772805'
  and created_on >0
  and persistent0_.REALM_ID='acme'
limit 1
```
by default the primary key index  is used:
![image](https://user-images.githubusercontent.com/314690/107161768-3aac0280-699f-11eb-94a5-cf8ea8a3d0f3.png)

with a more specific index:
```sql
create index idx_offline_uss_by_usersess on offline_user_session (offline_flag, realm_id, user_session_id);
```
the query is much quicker:
![image](https://user-images.githubusercontent.com/314690/107161717-e43ec400-699e-11eb-8f72-6ed0acf898fe.png)


The query `findUserSessionsByUserId` generates the following SQL:
```sql
--- PersistentUserSessionEntity.findUserSessionsByUserId

select persistent0_.OFFLINE_FLAG         as OFFLINE_1_41_,
       persistent0_.USER_SESSION_ID      as USER_SES2_41_,
       persistent0_.CREATED_ON           as CREATED_3_41_,
       persistent0_.DATA                 as DATA4_41_,
       persistent0_.LAST_SESSION_REFRESH as LAST_SES5_41_,
       persistent0_.REALM_ID             as REALM_ID6_41_,
       persistent0_.USER_ID              as USER_ID7_41_
from OFFLINE_USER_SESSION persistent0_
where persistent0_.OFFLINE_FLAG='1'
  and persistent0_.REALM_ID='acme'
  and persistent0_.USER_ID='87511de7-165c-4a34-891d-d0db60872644'
```
The uses a combined bitmap scan
![image](https://user-images.githubusercontent.com/314690/107161605-0d128980-699e-11eb-9824-50577937fb56.png)

If I create a more specific index:

```sql
create index idx_offline_uss_by_user on offline_user_session (offline_flag, realm_id, user_id);
```

Then the query is much quicker:
![image](https://user-images.githubusercontent.com/314690/107161624-2c111b80-699e-11eb-88a8-ee5f7e364a68.png)

thomasdarimont replied:
With this I'd say that the new queries are all covered by an index.

thomasdarimont replied:
I think the index definitions I choose are not optimal and need some additional tweaking.

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -270,8 +337,41 @@ private UserSessionEntity getUserSessionEntity(RealmModel realm, String id, bool
+            if (predicate.getBrokerUserId() != null) {
+                // TODO add support for offline user-session lookup by brokerUserId
+                // currently it is not possible to access the brokerUserId in offline user-session in a database agnostic way
+                throw new ModelException("Dynamic database lookup for offline user-sessions by brokerUserId is currently not supported.");
```
```suggestion
                throw new ModelException("Dynamic database lookup for offline user-sessions by brokerUserId is currently only supported for preloaded sessions.");
```

thomasdarimont replied:
Fixed

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -270,8 +337,41 @@ private UserSessionEntity getUserSessionEntity(RealmModel realm, String id, bool
+            if (predicate.getBrokerSessionId() != null) {
+                // TODO add support for offline user-session lookup by brokerSessionId
+                // currently it is not possible to access the brokerSessionId in offline user-session in a database agnostic way
+                throw new ModelException("Dynamic database lookup for offline user-sessions by brokerSessionId is currently not supported.");
```
```suggestion
                throw new ModelException("Dynamic database lookup for offline user-sessions by brokerSessionId is currently only supported for preloaded sessions.");
```

thomasdarimont replied:
Fixed

hmlnarik reviewed (changes requested):
Thank you @thomasdarimont for the changes.

Currently the OFFLINE_USER_SESSION and OFFLINE_CLIENT_SESSION lack the indices that are needed for index scan rather than seq scan of the tables in the introduced queries. Some indices would be needed. I've not gone through all the queries in detail, yet e.g. lookup by client_id in OFFLINE_CLIENT_SESSION needs to be backed by index, as well as lookup by realm.

To test the query performance and plans on Postgres, the following script might be useful that generates test data:
```sql
-- Generate 100 realms
insert into realm(id, name)
select md5(random()::text), md5(random()::text) from generate_series(1, 100) s;

-- Generate 1000 users in each realm
insert into user_entity(id, email_constraint, username, realm_id)
select md5(random()::text) || s.i, md5(random()::text), md5(random()::text), r.id 
  from realm r, generate_series(1, 1000) s(i);

-- generate persistent sessions for ~ 80 percent of all the users
INSERT INTO public.offline_user_session(
            user_session_id, user_id, realm_id, created_on, offline_flag, 
            data, last_session_refresh)
select
    md5(random()::text),
    u.id,
    u.realm_id,
    1,
    CASE WHEN random() > 0.5 then 'T' ELSE 'F' END,
    random()::text,
    (10000000*random())::int
from user_entity u TABLESAMPLE system (80);
```

thomasdarimont commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/stream/UserSessionPredicate.java:
```diff
@@ -108,6 +108,22 @@ public UserSessionPredicate brokerUserId(String id) {
+     * Returns the user id.
+     * @return
+     */
+    public String getUser() {
```
I just wanted to align the method name with the fieldname, which is `user`. I reverted the change.

thomasdarimont committed:
KEYCLOAK-11019 Fix exception message for unsupported brokered offline user-session lookups

thomasdarimont committed:
KEYCLOAK-11019 Fix naming in UserSessionPredicate

thomasdarimont committed:
KEYCLOAK-11019 Revise queries for offline user-session loading

Removed redundant condition checking.

We use the `coalesce` function to make the queries work for a
specific realm / client or all, e.g.:
```
clientSess.clientId = coalesce(CAST(:clientId as text), clientSess.clientId)
```
Note that the explicit cast is required to treat null values as text to avoid
confusing databases.

thomasdarimont committed:
KEYCLOAK-11019 Add indices to speed up lazy offline user-session lookups

thomasdarimont commented:
I revised the code and rebased on current master. I'll be working on the required tests this week.

thomasdarimont commented:
Just rebased on current master

hmlnarik reviewed (changes requested):
Hi @thomasdarimont Thank you for patience - it took a while for me to get to this PR as I'm still in a catch-up mode after an unexpected long term leave.
Overall it looks good to me, only the indices could be improved as you have also mentioned. I've done some analysis of the queries [here](https://docs.google.com/spreadsheets/d/14nk2i7bG0dCeI0MLqANoxEXX1VIehyfOZ5qFQIXoCrs/edit?usp=sharing). Based on the analysis, I am suggesting change the indices, please see inline.

hmlnarik commented:
Hi @thomasdarimont, would you have time to move this PR forward?

hmlnarik commented:
Hi @thomasdarimont This PR has been inactive for almost two weeks. Will you be able to move this PR forward?

thomasdarimont commented:
@hmlnarik yes, I have a few days vacation ahead and I'm going to use some of my time to work on this.

As I've been also involved in the user-profile / validation API discussion, I had to adjust some priorities ;-)

thomasdarimont commented on model/jpa/src/main/resources/META-INF/jpa-changelog-13.0.0.xml:
```diff
@@ -38,4 +38,23 @@
+            <column name="CLIENT_ID" type="VARCHAR(36)"/>
+        </createIndex>
+
+        <createIndex tableName="OFFLINE_USER_SESSION" indexName="IDX_OFFLINE_USS_BY_USER">
+            <column name="OFFLINE_FLAG" type="VARCHAR(4)"/>
+            <column name="REALM_ID" type="VARCHAR(36)"/>
+            <column name="USER_ID" type="VARCHAR(36)"/>
+        </createIndex>
+
+        <createIndex tableName="OFFLINE_USER_SESSION" indexName="IDX_OFFLINE_USS_BY_USERSESS">
+            <column name="OFFLINE_FLAG" type="VARCHAR(4)"/>
+            <column name="REALM_ID" type="VARCHAR(36)"/>
+            <column name="USER_SESSION_ID" type="VARCHAR(36)"/>
+        </createIndex>
```
Thanks, I'll adjust the indices as you suggested.

thomasdarimont committed:
KEYCLOAK-11019 Revise indices according to analysis from code review

thomasdarimont commented:
@hmlnarik I updated the indices and rebased the branch again. 

Btw. did you see #7902? I think that and this PR would work well together.

There is also the idea of expiring offline session cache entries to keep memory consumption low mentioned here: https://github.com/keycloak/keycloak/pull/7902#issuecomment-814794402

hmlnarik commented:
@thomasdarimont Apologies I did not respond. Thank you for the updates, the PR looks much better. Could you please have a look at the failures of GitHub actions? They are related to this PR.

Yup, #7902 looks interesting though I've not yet have a chance to go through it in detail.

Flintholm commented:
@thomasdarimont We (me and @nigtrifork) have been experimenting with lazy loading of the offlineClientSessions as an extension of this PR. The idea was to avoid the synchonization issues mentionen in the discoussion in #7902 if the offline session caches are configured to expire.
Do you think this could be viable solution for basically keeping the offline sessions in db, and only use the caches to most recent sessions ?

martin-kanis commented:
@thomasdarimont I opened a new PR #8033 which includes this PR plus rebase to current master plus adds some model tests. Let us please know whether you wish pull those changes here or we can close this PR and proceed with #8033. Thank you.

thomasdarimont commented:
@martin-kanis thanks for the update! I have no problem closing this PR and continue in the combined efforts in the new one.

hmlnarik commented:
Thank you @thomasdarimont and @martin-kanis. Closing this one in favor of #8033.
